wasm/src/lib.rs

```diff
@@ -0,0 +1,211 @@
+        #[cfg(not(test))]
+        pub(super) fn execute_query(ptr: WasmUsize, len: WasmUsize) -> WasmQueryResult;
+
+        /// Executes encoded instruction by providing offset and length
+        /// into WebAssembly's linear memory where instruction is stored
+        ///
+        /// # Warning
+        ///
+        /// This function doesn't take ownership of the provided allocation
+        /// but it does transfer ownership of the result to the caller
+        #[cfg(not(test))]
+        pub(super) fn execute_instruction(ptr: WasmUsize, len: WasmUsize);
+    }
+}
+
+/// Decode the object from given pointer and length
+///
+/// # Warning
+///
+/// This method takes ownership of the given pointer
+///
+/// # Safety
+///
+/// It's safe to call this function as long as it's safe to construct `Box<[u8]>` from the given pointer
+// `WasmUsize` is always pointer sized
+#[allow(clippy::cast_possible_truncation)]
+pub unsafe fn _decode_from_raw<T: Decode>(ptr: WasmUsize, len: WasmUsize) -> T {
+    let bytes = Box::from_raw(core::slice::from_raw_parts_mut(ptr as *mut _, len as usize));
+
+    #[allow(clippy::expect_used, clippy::expect_fun_call)]
+    T::decode(&mut &bytes[..]).expect(
+        format!(
+            "Decoding of {} failed. This is a bug",
+            core::any::type_name::<T>()
+        )
+        .as_str(),
+    )
+}
+
+/// Encode the given object and call the given function with the pointer and length of the allocation
+///
+/// # Warning
+///
+/// Ownership of the returned allocation is transfered to the caller
+///
+/// # Safety
+///
+/// The given function must not take ownership of the pointer argument
+unsafe fn encode_and_execute<T: Encode, O>(
+    obj: T,
+    fun: unsafe extern "C" fn(WasmUsize, WasmUsize) -> O,
+) -> O {
+    // NOTE: It's imperative that encoded object is stored on the heap
+    // because heap corresponds to linear memory when compiled to wasm
+    let bytes = obj.encode();
+
+    // `WasmUsize` is always pointer sized
+    #[allow(clippy::cast_possible_truncation)]
+    let ptr = bytes.as_ptr() as WasmUsize;
+    // `WasmUsize` is always pointer sized
+    #[allow(clippy::cast_possible_truncation)]
+    let len = bytes.len() as WasmUsize;
+
+    fun(ptr, len)
+}
+
+/// Executes the given query on the host environment
+pub fn execute_query(query: QueryBox) -> QueryResult {
+    #[cfg(not(test))]
+    use host::execute_query as host_execute_query;
+    #[cfg(test)]
+    use tests::_iroha_wasm_execute_query_mock as host_execute_query;
+
+    // Safety: - `host_execute_query` doesn't take ownership of it's pointer parameter
+    //         - ownership of the returned result is transfered into `_decode_from_raw`
+    unsafe {
+        let host::WasmQueryResult(res_ptr, res_len) = encode_and_execute(query, host_execute_query);
+        _decode_from_raw(res_ptr, res_len)
+    }
+}
+
+/// Execute the given instruction on the host environment
+pub fn execute_instruction(instruction: Instruction) {
```

Comment:
we can have only one generic execute method instead of `execute_instruction` and `execute_query`. Would that be preferred?

Reply:
I would. We already have `execute` for both impls of `ValidQuery` and `Execute`. While having separate `execute_instruction` will sometimes be clearer, I don't see why the same clarity can't be achieved using `Execute::execute(instruction)`. At the same time, in most cases it's obvious whether or not what you're dealing with is an instruction or a query.

Reply:
changed, check the latest commit